### PR TITLE
Fix using default context storage implemenation in LazyStorage

### PR DESCRIPTION
**Pull request by dengliming**
fixes [https://github.com/open-telemetry/opentelemetry-java/issues/1819](https://github.com/open-telemetry/opentelemetry-java/issues/1819)

### CODE REVIEW

**Review by anuraaga — COMMENTED** [review]
Thanks a lot for the PR! I think we do want to be able to use a context provider implicitly if it's the only one since it's fairly unambiguous - the issue was that the property may also be set and to a different provider so we should handle that case specially. Does that make sense?

**Comment by dengliming** [comment]
@anuraaga So The `providerClassName` is not required, is that right? I'm still a little confused 

> I think we do want to be able to use a context provider implicitly if it's the only one since it's fairly unambiguous

Do you mean that the `io.opentelemetry.context.contextStorageProvider ` system property is not set and there is exactly a single provider found by `ServiceLoader.load(ContextStorageProvider.class)`?

**Comment by anuraaga** [comment]
Right if it's unset we can use the single implementation. We want to prevent a user explicitly wanting an implementation by setting the property but getting a different one.

**Review by jkwatson — COMMENTED** [review]
Is it possible to have unit tests to cover this change?

**Comment by dengliming** [comment]
@jkwatson yeah. Seems the unit tests only cover `ContextStorage.get()`. We don't have unit tests for `LazyStorage` now. I will add one later if that make senses.

File: context/src/main/java/io/opentelemetry/context/LazyStorage.java [hunk 1]
```diff
@@ -48,17 +48,11 @@ private static ContextStorage createStorage(AtomicReference<Throwable> deferredS
...
     }
 
-    if (providers.size() == 1) {
-      ContextStorageProvider provider = providers.get(0);
-      try {
-        return provider.get();
-      } catch (Throwable t) {
-        deferredStorageFailure.set(t);
-        return DefaultContext.threadLocalStorage();
+    if (providerClassName.isEmpty()) {
```
Comment (anuraaga) [inline context/src/main/java/io/opentelemetry/context/LazyStorage.java:51]: @dengliming Sorry if my suggestions weren't clear, thanks for bearing with me. We have these success cases

- 1 provider, providerClassName empty
- 1 provider, providerClassName matches the provider <- currently missing
- multiple providers, providerClassName matches one of them

Hope this makes it clear

**Review by anuraaga — COMMENTED** [review]
Yeah since this is SPI I think tests are more complicated by needing a testset per case. I'm ok with a followup. Thanks for the help @dengliming

File: context/src/main/java/io/opentelemetry/context/LazyStorage.java (hunk 1 above)
Comment (dengliming) [reply context/src/main/java/io/opentelemetry/context/LazyStorage.java:51]: yeah. a little confused. Can you also add the right return for each case？

> 1 provider, providerClassName matches the provider <- currently missing

don't know what is missing. The following code should cover this case. is that correct?
```
for (ContextStorageProvider provider : providers) {
      if (provider.getClass().getName().equals(providerClassName)) {
        return provider.get();
      }
    }
```

Comment (anuraaga) [reply context/src/main/java/io/opentelemetry/context/LazyStorage.java:51]: Sure updated my comment, in every other case we log and return default storage.

Yeah starting with that, and than one more branch for if there's a single provider and empty providerClassName seems fine to me.

Comment (dengliming) [reply context/src/main/java/io/opentelemetry/context/LazyStorage.java:51]: I think the current changes can satisfy all the cases. Just need log in some case. WDYT?

Comment (anuraaga) [reply context/src/main/java/io/opentelemetry/context/LazyStorage.java:51]: Sorry got it! Yeah seems good, really needed to expand the diff

**Review by anuraaga — APPROVED** [review]

**Review by thisthat — APPROVED** [review]

**Pull request merged**